### PR TITLE
Refactoring on xpdAcq

### PR DESCRIPTION
**Pull request by chiahaoliu**
@CJ-Wright @sbillinge  @danielballan

First, I need to apologize for such a long PR and have you guys reviewed; it came from many historical issues. 

This PR consistent of refactorings on xpdAcq, main features of this PR are:
  1. turn ``prun`` into a customized run engine that allows specified logic of taking dark frame every time before taking light frame
  2. make ``ct``, ``tseries`` and ``Tramp`` into bluesky plans 
  3.  implement customized ``validated yaml dict`` for beamtime objects so that inherited contents and yaml files will be automatically updated when parents are modified.
  4. add ``pyFAI`` into calibration and integration workflow.
  5. scattered updates on attributes of ``Glbl`` class to suit current workflow.

bullet points 1 to 3 thanks to huge help from @danielballan.

Code itself passed tests in simulation and unittest, but there is alway a room for better implementation details and suggestions are welcome.

Docstring and corresponding documentations are highly incomplete at this point (I am working on it ), so feel free to tag me when you found function signatures are unclear. 

Note: 
to set up a simulation environment, please just create an ipython profile with ``999-load.py`` inside ``startup/`` of this profile.

### CODE REVIEW

**Comment by sbillinge** [comment]
Wait, taking dark before every light is the wrong behavior....

File: setup.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 setup(
     name = 'xpdacq',
-    version = '1.0.2',
```
Comment (danielballan) [inline setup.py]: You probably should not decrease the version. Better to call this 1.4.

File: xpdacq/xpdacq.py [hunk 2]
```diff
@@ -1,823 +1,416 @@
...
+def periodic_dark(plan):
+    """
+    a plan wrapper that takes a plan and inserts `take_dark`
+
+    The `take_dark` plan is inserted on the fly before the beginning of
+    any new run after a period of time defined by `glbl.dk_window` has passed.
+    """
+    need_dark = True
+
+    def insert_take_dark(msg):
```
Comment (danielballan) [inline xpdacq/xpdacq.py]: This magic makes me really happy. :- )

File: setup.py (hunk 1 above)
Comment (sbillinge) [inline setup.py]: ouch, I don't want to go to v > 1.  Is 1.0.2 in a release tag?

File: xpdacq/xpdacq.py (hunk 2 above)
Comment (sbillinge) [inline xpdacq/xpdacq.py]: ah, ok.  Thanks. It makes me happy too.  Is this the default behavior in Acq?

File: xpdacq/tests/test_prun.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        self.assertTrue('sc_calibration_md' in open_run)
+        self.assertEqual(open_run['sc_calibration_md'],
+                         re_auto_calibration_md_dict)
+
+    def test_open_collection(self):
+        # no collection
+        self.assertRaise(RuntimeError, lambda: prun(0,0))
+        # collection num
+        open_collection('unittest')
+        self.aseertEqual(glbl.collection, [])
```
Comment (danielballan) [inline xpdacq/tests/test_prun.py]: typo

File: xpdacq/Ni24.D [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Calibrant: Nickel (Ni)
```
Comment (CJ-Wright) [inline xpdacq/Ni24.D]: Would it be possible to put this in a subfolder? At some point you might want other calibration standards and you might not want them at the top level folder.

File: xpdacq/analysis.py [hunk 5]
```diff
@@ -154,16 +154,18 @@ def save_tiff(headers, dark_subtraction=True, *, max_count=None):
             try:
                 # bluesky only looks for uid it defines
                 dark_search = {'group': 'XPD',
-                               'sc_dark_uid': dark_uid_appended} # the one we need to look up data
+                               'uid': dark_uid_appended} # the one we need to look up for dark
```
Comment (CJ-Wright) [inline xpdacq/analysis.py]: Can you just use `header.start['sc_dk_field_uid']` here, or do you need it again later?

Comment (CJ-Wright) [inline xpdacq/analysis.py]: Also would it make sense to put the `try`  at line 160?

File: setup.py (hunk 1 above)
Comment (danielballan) [inline setup.py]: Fair enough. Your git tags look to be in good shape (latest is 0.3) so I would make this 0.4.0 (or something).

Comment (danielballan) [inline setup.py]: I should add: we love using "versioneer," which auto-detects your latest git tag and makes `__version__` for you. Once that's set up, you never have to think about it again. I'll help someone set that up in exchange for more bluesky docs reviews.... ;- )

File: xpdacq/beamtime.py [hunk 6]
```diff
@@ -1,720 +1,408 @@
...
+    computed_exposure = num_frame*acq_time
+    glbl.area_det.images_per_set.put(num_frame)
+    # print exposure time
+    print("INFO: requested exposure time = {} - > computed exposure time"
+          "= {}".format(exposure, computed_exposure))
+    return (num_frame, acq_time, computed_exposure)
+
+
+def ct(dets, exposure, *, md=None):
+    pe1c, = dets
```
Comment (CJ-Wright) [inline xpdacq/beamtime.py]: Would it be worthwhile to add some docs here?

File: xpdacq/beamtime.py [hunk 7]
```diff
@@ -1,720 +1,408 @@
...
+                        'sp_computed_exposure': computed_exposure,
+                        'sp_type': 'ct',
+                        # need a name that shows all parameters values
+                        #'sp_name': 'ct_<exposure_time>',
+                        'sp_uid': str(uuid.uuid4()),
+                        'sp_plan_name': 'ct'})
+    plan = bp.count([glbl.area_det], md = _md)
+    plan = bp.subs_wrapper(plan, LiveTable([glbl.area_det]))
+    yield from plan
+
```
Comment (CJ-Wright) [inline xpdacq/beamtime.py]: PEP8: two lines between functions

File: xpdacq/xpdacq.py (hunk 2 above)
Comment (danielballan) [inline xpdacq/xpdacq.py]: Yes, this is the default. The so-called "dark_strategy" argument in `CustomizedRunEngine.__call__` makes this configurable at *execution time*, separate from the rest of the logic of the plan. That's a little different that bluesky's RunEngine, but it makes sense for your workflow, and it matches the way the original xpdAcq injected this option.

File: xpdacq/beamtime.py [hunk 8]
```diff
@@ -1,720 +1,408 @@
...
+                        # need a name that shows all parameters values
+                        # 'sp_name': 'tseries_<exposure_time>',
+                        'sp_uid': str(uuid.uuid4()),
+                        'sp_plan_name': 'tseries'})
+    plan = bp.count([glbl.area_det], num, delay, md=_md)
+    plan = bp.subs_wrapper(plan, LiveTable([glbl.area_det]))
+    yield from plan
+
+def _nstep(start, stop, step_size):
+    ''' return (start, stop, nsteps)'''
```
Comment (CJ-Wright) [inline xpdacq/beamtime.py]: This return is different from the actual returned values.

File: xpdacq/beamtimeSetup.py [hunk 9]
```diff
@@ -64,42 +174,48 @@ def _end_beamtime(base_dir=None,archive_dir=None,bto=None, usr_confirm = 'y'):
...
+
+
+def _load_bt_info(bt_obj, required_fields):
+    # grab information
+    bt_info_list = []
+    for el in required_fields:
+        #print('loaded bt info = {}'.format(dict(bt_obj)))
+        bt_info = bt_obj.get(el)
+        if bt_info is None:
+            print("WARNING: required beamtime information {} doesn't exit."
```
Comment (CJ-Wright) [inline xpdacq/beamtimeSetup.py]: do you mean `doesn't exist`?

File: xpdacq/Ni24.D (hunk 4 above)
Comment (chiahaoliu) [inline xpdacq/Ni24.D]: Yes, that was the implementation detail I didn't fully cover at that time.

Btw, this is a customized d-spacing file made by @sbanerjee23 which capture additional rings. Just want to nominate him :p

File: setup.py (hunk 1 above)
Comment (chiahaoliu) [inline setup.py]: you already know who just raised his hand :p

File: xpdacq/analysis.py (hunk 5 above)
Comment (chiahaoliu) [inline xpdacq/analysis.py]: This file is deprecated. Full functionality is moved to ``xpdAn``. I will check if the same error happen in there.

File: xpdacq/Ni24.D (hunk 4 above)
Comment (sbanerjee23) [inline xpdacq/Ni24.D]: Haha it's just a minor edit to our trusty Ni03.cal d-spacings from fit2d, but has more spacings than the Ni.D deafault in the PyFAI library. This just helps with refining the detector non-orthogonality.

File: xpdacq/beamtime.py (hunk 8 above)
Comment (chiahaoliu) [inline xpdacq/beamtime.py]: updated in the new commits

File: xpdacq/beamtimeSetup.py (hunk 9 above)
Comment (chiahaoliu) [inline xpdacq/beamtimeSetup.py]: updated in the new commits

File: xpdacq/beamtime.py (hunk 7 above)
Comment (chiahaoliu) [inline xpdacq/beamtime.py]: updated in my new commits

File: xpdacq/tests/test_prun.py (hunk 3 above)
Comment (chiahaoliu) [inline xpdacq/tests/test_prun.py]: ``unittest`` is updated in the new commits

File: xpdacq/beamtime.py (hunk 6 above)
Comment (chiahaoliu) [inline xpdacq/beamtime.py]: Yes, this is for sure. However, I would prefer to do it in the next PR, a PR focus on docs and web page

File: xpdacq/Ni24.D (hunk 4 above)
Comment (danielballan) [inline xpdacq/Ni24.D]: Could this be submitted as a patch to PyFAI itself?

File: xpdacq/beamtime.py (hunk 6 above)
Comment (danielballan) [inline xpdacq/beamtime.py]: When you get around to that (and I agree it's reasonable to do it in a separate pass, given the overhaul nature of this PR) you can save time by copy/pasting the bluesky count docstring (etc.) and editing as needed.

File: xpdacq/Ni24.D (hunk 4 above)
Comment (chiahaoliu) [inline xpdacq/Ni24.D]: @sbanerjee23  could you do this (submit as a patch to pyFAI)? Take entire credit :)

**Comment by CJ-Wright** [comment]
Would it be possible to merge this so that we can have multiple branches/forks for working on this? This way I can help with this release.

**Comment by sbillinge** [comment]
The branch now has conflicts.  @chiahaoliu , can you resolve those and we can merge it, so we can move on!  hehe.

@CJ-Wright I will try and add you as an admin.  Wish me luck.

**Comment by chiahaoliu** [comment]
@sbillinge conflicts have been resolved. They come from outdated files. Code still passes unittest and simulation. Thanks for pulling in :)

**Pull request merged**